desc/objectives/utils.py

```diff
@@ -237,95 +246,63 @@ def factorize_linear_constraints(constraints, objective_args):  # noqa: C901
-            b_ = -obj.compute_scaled(jnp.zeros(obj.dimensions[arg]))
-            if arg not in A.keys():
-                A[arg] = A_
-
-                _A_unweighted[arg] = (
-                    obj.normalization * (A_.T / obj.weight).T
-                )  # undo normalization here for each indiv A
-                b[arg] = b_
-                unfixed_args.append(arg)
-                arg_obj_lists[arg] = [obj]
-            elif b_.size > 0:  # we want to stack these vertically if the same arg
-                # but only if the constraint actually constrains anything
-                # if not, i.e. if attempted to fix a mode that is not in the basis,
-                # b_ is an empty array and does not need to be added
-                rk_before = jnp.linalg.matrix_rank(A[arg])
-                A[arg] = jnp.vstack((A[arg], A_))
-                rk_after = jnp.linalg.matrix_rank(A[arg])
-                if (
-                    rk_after < rk_before + A_.shape[0]
-                ):  # if less, there are some lin dependent rows of A_ and A[arg]
-                    # for example, could be trying to constrain same mode with 2
-                    # different values
-                    raise RuntimeError(
-                        f"Incompatible constraints given for {arg}!"
-                        + f" constraint {obj} is incompatible with"
-                        + f" {*arg_obj_lists[arg],}"
-                    )
-                _A_unweighted[arg] = jnp.vstack(
-                    (A[arg], A_ / obj.weight * obj.normalization)
-                )
-                arg_obj_lists[arg] += [obj]
-
-                b[arg] = jnp.hstack((b[arg], b_))
-    # find inverse of the now-combined constraint matrices for each arg
-    for key in list(A.keys()):
-        if A[key].shape[0]:
-            Ainv[key], Z_ = svd_inv_null(_A_unweighted[key])
-        else:
-            Ainv[key] = _A_unweighted[
-                key
-            ].T  # make inverse matrices using already unweighted A
-
-    # catch any arguments that are not constrained
-    for arg in x_idx.keys():
-        if arg not in constraint_args:
-            unfixed_args.append(arg)
-            A[arg] = jnp.zeros((1, dimensions[arg]))
-            b[arg] = jnp.zeros((1,))
-
-    # full A matrix for all unfixed constraints
-    if len(A):
-        unfixed_idx = jnp.concatenate(
-            [x_idx[arg] for arg in arg_order if arg in A.keys()]
+        A_ = {
+            arg: obj.derivatives["jac"][arg](
+                *[jnp.zeros(obj.dimensions[arg]) for arg in obj.args]
+            )
+            for arg in args
+        }
+        # using obj.compute instead of obj.target to allow for correct scale/weight
+        b_ = -obj.compute_scaled(*[jnp.zeros(obj.dimensions[arg]) for arg in obj.args])
+        A.append(A_)
+        b.append(b_)
+
+    A_full = jnp.vstack([jnp.hstack([Ai[arg] for arg in args]) for Ai in A])
```

Comment:
So here, A contains the dict which has A matrices for each arg, and Ai[arg] is empty(?) if the objective does not rely on that arg? so then when the inner jnp.hstack happens, it creates an A matrix for each objective, then  the vstack puts the A matrices for all the objectives together?

Reply:
Not empty, it will just be a matrix of zeros of the appropriate size if that constraint doesn't take that arg.

Reply:
but otherwise yes

Reply:
`A` itself is a list, one for each constraint, and each element of that is a dictionary of the jacobian wrt each arg